resourcemanager/pool/spool/spool.go

```diff
@@ -0,0 +1,171 @@
+// limitations under the License.
+
+package spool
+
+import (
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/pingcap/tidb/metrics"
+	"github.com/pingcap/tidb/resourcemanager"
+	"github.com/pingcap/tidb/resourcemanager/pool"
+	"github.com/pingcap/tidb/resourcemanager/pooltask"
+	"github.com/pingcap/tidb/resourcemanager/util"
+	"github.com/pingcap/tidb/util/logutil"
+	"github.com/pingcap/tidb/util/mathutil"
+	"github.com/prometheus/client_golang/prometheus"
+	"go.uber.org/zap"
+)
+
+// Pool is a single goroutine pool. it can not reuse the goroutine.
+type Pool struct {
+	wg                 sync.WaitGroup
+	mu                 sync.Mutex
+	options            *Options
+	capacity           atomic.Int32
+	running            atomic.Int32
+	isStop             atomic.Bool
+	concurrencyMetrics prometheus.Gauge
+	taskManager        pooltask.TaskManager[any, any, any, any, pooltask.NilContext]
+	pool.BasePool
+}
+
+// NewPool create a single goroutine pool.
+func NewPool(name string, size int32, component util.Component, options ...Option) (*Pool, error) {
+	opts := loadOptions(options...)
+	result := &Pool{
+		options:            opts,
+		concurrencyMetrics: metrics.PoolConcurrencyCounter.WithLabelValues(name),
+		taskManager:        pooltask.NewTaskManager[any, any, any, any, pooltask.NilContext](size), // TODO: this general type
+	}
+	if size == 0 {
+		return nil, pool.ErrPoolParamsInvalid
+	}
+	result.SetName(name)
+	result.capacity.Store(size)
+	result.concurrencyMetrics.Set(float64(size))
+	err := resourcemanager.InstanceResourceManager.Register(result, name, component)
+	if err != nil {
+		return nil, err
+	}
+	return result, nil
+}
+
+// Tune changes the pool size.
+func (p *Pool) Tune(size int) {
+	if size == 0 {
+		return
+	}
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	p.SetLastTuneTs(time.Now())
+	p.capacity.Store(int32(size))
+	p.concurrencyMetrics.Set(float64(size))
+}
+
+// Run runs a function in the pool.
+func (p *Pool) Run(fn func()) error {
+	if p.isStop.Load() {
+		return pool.ErrPoolClosed
+	}
+	_, run := p.checkAndAddRunning(1)
+	if !run {
+		return pool.ErrPoolOverload
+	}
+	p.taskManager.RegisterTask(p.GenTaskID(), 1)
+	p.run(fn)
+	return nil
+}
+
+// Cap returns the capacity of the pool.
+func (p *Pool) Cap() int {
+	return int(p.capacity.Load())
+}
+
+// Running returns the number of running goroutines.
+func (p *Pool) Running() int {
+	return int(p.running.Load())
+}
+
+func (p *Pool) run(fn func()) {
+	p.wg.Add(1)
+	go func() {
+		defer func() {
+			if r := recover(); r != nil {
+				logutil.BgLogger().Error("recover panic", zap.Any("recover", r))
+			}
+			p.wg.Done()
+			p.running.Add(-1)
+		}()
+		fn()
+	}()
+}
+
+// RunWithConcurrency runs a function in the pool with concurrency.
+func (p *Pool) RunWithConcurrency(fns chan func(), concurrency uint32) error {
+	if p.isStop.Load() {
+		return pool.ErrPoolClosed
+	}
+	conc, run := p.checkAndAddRunning(concurrency)
```

Comment:
That's a little different from my requirement, it books as many goroutines as we required in the pool, I may call this function like:

```go
maxConcurrency := 15
queue := make(chan func(), maxConcurrency)

// I don't know the concurrency required when spawning the tasks.
p.RunWithConcurrency(queue, maxConcurrency)

go func() {
	for {
		// build tasks here
		queue <- func() { ... }
	}
} ()
```

So that it may occupy extra concurrency than its requirement and open useless goroutines here.